Source/RuntimeImageLoader/Private/Helpers/GIFLoader.cpp

```diff
@@ -0,0 +1,217 @@
+	}
+
+	/* clean up */
+	FLibnsgifHandler::FunctionPointerNsgifDestroy()(Gif);
+	free(Data);
+}
+
+int32 FRuntimeGIFLoaderHelper::GetWidth() const
+{
+	return FLibnsgifHandler::FunctionPointerNsgifGetInfo()(Gif)->width;
+}
+
+int32 FRuntimeGIFLoaderHelper::GetHeight() const
+{
+	return FLibnsgifHandler::FunctionPointerNsgifGetInfo()(Gif)->height;
+}
+
+int32 FRuntimeGIFLoaderHelper::GetTotalFrames() const
+{
+	return FLibnsgifHandler::FunctionPointerNsgifGetInfo()(Gif)->frame_count;
+}
+
+void FRuntimeGIFLoaderHelper::Decode(nsgif_t* gif, bool first)
+{
+	nsgif_error err;
+	uint32_t frame_prev = 0;
+	const nsgif_info_t* info;
+	info = FLibnsgifHandler::FunctionPointerNsgifGetInfo()(gif);
+
+	// Calculate the total number of pixels (frame_count * width * height)
+	int32 TotalPixels = info->frame_count * info->width * info->height;
+	TextureData.Empty(TotalPixels);
+	TextureData.AddUninitialized(TotalPixels);
+
+	// Decode the frames
+	while (true) {
+		nsgif_bitmap_t* bitmap;
+		const uint8* image;
+		uint32_t frame_new;
+		uint32_t delay_cs;
+		nsgif_rect_t area;
+
+		err = FLibnsgifHandler::FunctionPointerNsgifFramePrepare()(gif, &area, &delay_cs, &frame_new);
+		if (err != NSGIF_OK) {
+			Warning("nsgif_frame_prepare", err);
+			return;
+		}
+
+		if (frame_new < frame_prev) {
+			// Must be an animation that loops. We only care about
+			// decoding each frame once in this utility.
+			return;
+		}
+		frame_prev = frame_new;
+
+		err = FLibnsgifHandler::FunctionPointerNsgifFrameDecode()(gif, frame_new, &bitmap);
+		if (err != NSGIF_OK) {
+			// Continue decoding the rest of the frames.
+		}
+		else {
+			image = (const uint8*)bitmap;
+			for (uint32_t i = 0; i < info->height * info->width; i++) {
+				uint32_t y = i / info->width;
+				uint32_t x = i % info->width;
+
+				size_t z = (y * info->width + x) * 4;
+				uint8 Red = image[z];
+				uint8 Green = image[z + 1];
+				uint8 Blue = image[z + 2];
+				uint8 Alpha = image[z + 3];
+
+				int32 PixelIndex = ((frame_new * info->height + y) * info->width + x);
+				TextureData[PixelIndex] = FColor(Red, Green, Blue, Alpha);
+			}
+		}
+
+		if (delay_cs == NSGIF_INFINITE) {
+			// This frame is the last.
+			return;
+		}
+	}
+}
+
+void FRuntimeGIFLoaderHelper::GetNextFrame(TArray<FColor>& NextFramePixels, int32 FrameIndex)
+{
+	if (FrameIndex > GetTotalFrames() - 1) FrameIndex = 0;
+	// Calculate the starting index of the desired frame in the OutPixels array
+	int32 StartIndex = FrameIndex * GetWidth() * GetHeight();
+
+	// Resize the SingleFramePixels array to hold the pixel data for the single frame
+	NextFramePixels.Empty(GetFramePixels());
+	NextFramePixels.AddUninitialized(GetFramePixels());
+
+	// Copy the pixel data for the single frame into SingleFramePixels
+	for (int32 i = 0; i < GetFramePixels(); i++)
+	{
+		int32 index = StartIndex + i;
+
+		if (index >= 0 && index < TextureData.Num())
+		{
+			NextFramePixels[i] = TextureData[index];
```

Comment:
Btw, instead of filling NextFramePixels with pixels values (and wasting CPU time) you can simply return pointer to some place in TextureData i.e. 

`&TextureData[StartIndex]`

Let me know what you think. You don't really need to use an additional array here i.e. NextFramePixels as you can always return `const uint8*`

Reply:
Resolved